staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go

```diff
@@ -142,6 +142,17 @@ func (a *Webhook) ValidateInitialization() error {
 // or an error if an error was encountered during evaluation.
 func (a *Webhook) ShouldCallHook(h webhook.WebhookAccessor, attr admission.Attributes, o admission.ObjectInterfaces) (*WebhookInvocation, *apierrors.StatusError) {
 	var err *apierrors.StatusError
+
+	matches, err := a.namespaceMatcher.MatchNamespaceSelector(h, attr)
+	if !matches || err != nil {
+		return nil, err
+	}
```

Comment:
This puts a namespace lookup (potentially a live network request) and several error paths ahead of checking if the webhook's rules even apply to this request. I'd like to see benchmark evidence that this is actually cheaper, and consider how to avoid returning an error if the webhook's rules don't even apply to the request.

Reply:
Updated, result:
BenchmarkShouldCallHook-4         329312              3182 ns/op
BenchmarkShouldCallHook-4         510973              2305 ns/op

Reply:
> how to avoid returning an error if the webhook's rules don't even apply to the request.

I have a question: where is the error happened, when we build invocation WebhookAccessor and Attributes won't change, right?

Reply:
Updated benchmark, make NewValidatingWebhookAccessor outside loop, result is
BenchmarkShouldCallHook-4        1329668               902 ns/op             288 B/op          6 allocs/op
BenchmarkShouldCallHook-4        8379460               140 ns/op              32 B/op          2 allocs/op

@liggitt  Do you think this bench test reasonable?

Reply:
> consider how to avoid returning an error if the webhook's rules don't even apply to the request.

Do you mean MatchNamespaceSelector returns an error while invocation(rules don't even apply to the request) is nil? We want to continue next hook but returned an error. Yeah, I haven't caught this scenario.

This just hide the error(mostly a get request, other occurred error might be logic bug) when no rules applied to request, once rules applied, error would happen, right?

Reply:
> Do you mean MatchNamespaceSelector returns an error while invocation(rules don't even apply to the request) is nil? We want to continue next hook but returned an error. Yeah, I haven't caught this scenario.

Exactly. We should not return an error here for webhooks which do not even apply to the incoming request.

Reply:
> This puts a namespace lookup (potentially a live network request) ... ahead of checking if the webhook's rules even apply to this request.

this is still outstanding as well

Reply:
Yeah, it can't avoid returning an error if the webhook's rules don't even apply to the request, so I close it.

Reply:
> Exactly. We should not return an error here for webhooks which do not even apply to the incoming request.

@liggitt addressed it. PTAL.

Reply:
Sorry, it's not right. it will break https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L175